File: java/hello2/app/src/main/java/com/biz/hello2/MainActivity.java
package com.biz.hello2;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.snackbar.Snackbar;

public class MainActivity extends AppCompatActivity {

    private TextView txtView1;
    private TextView txtView2;
    private Button btnClick1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        txtView1 = findViewById(R.id.txtView1);
        txtView2 = findViewById(R.id.txtView2);
        btnClick1 = findViewById(R.id.btnClick);

        txtView1.setText("나는 TextView1 입니다.");
        txtView2.setText("나는 TextView2 입니다.");

        View.OnClickListener btnListener = new BtnClickListener();
        btnClick1.setOnClickListener(btnListener);

        btnClick1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(MainActivity.this, "반갑다", Toast.LENGTH_LONG).show();
                txtView2.setText("음 바꿀꺼야~~");
                Snackbar.make(v, "나는 쓰렉바입니다", Snackbar.LENGTH_LONG).show();
            }
        });


        // xml 에 설정된 컴포넌트를 java code에서 사용하기 위해서
        // 연결하는 작업
        Log.d("onCreate", "onCreate: " + txtView1.getText());
    }
}

File: kotlin/TodoList/app/src/main/java/github/sun5066/MainActivity.kt
package github.sun5066

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputEditText
import github.sun5066.adapter.TodoViewAdapter
import github.sun5066.adapter.TodoViewModel
import github.sun5066.model.TodoVO
import java.text.SimpleDateFormat

class MainActivity : AppCompatActivity() {

    private lateinit var txtTodoInput: TextInputEditText
    private lateinit var btnAdd: Button
    private lateinit var todoViewModel: TodoViewModel
    private lateinit var todoViewAdapter: TodoViewAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        txtTodoInput = findViewById(R.id.input_todo)
        btnAdd = findViewById(R.id.btn_add)
        todoViewModel = TodoViewModel(this.application)

        txtTodoInput.setOnEditorActionListener { view, actionId, event ->
            return@setOnEditorActionListener when (actionId) {
                EditorInfo.IME_ACTION_SEND -> {
                    this.todoAdd(view)
                    true
                }
                else -> false
            }
        }

        btnAdd.setOnClickListener { view ->
            this.todoAdd(view)
        }

        var todoList: MutableList<TodoVO> = mutableListOf()
        todoViewAdapter = TodoViewAdapter(
            todoList,
            onDelete = { id -> todoViewModel.delete(id as Long) },
            onUpdate = { id -> this.todoUpdate(id as Long) },
            onComplete = { id -> this.todoComplete(id as Long) }
        )

        todoViewModel =
            ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory(application))
                .get(TodoViewModel::class.java)

        todoViewModel.selectAll().observe(this, {
            it?.let { todoViewAdapter.setList(it) }
            todoViewAdapter.notifyDataSetChanged()
        })

        val recyclerView: RecyclerView = findViewById(R.id.data_list_view)
        recyclerView.adapter = todoViewAdapter

        val layoutManager = LinearLayoutManager(this)
        recyclerView.layoutManager = layoutManager
    }

    fun todoAdd(view: View) {
        val txtTodo = txtTodoInput.text.toString()

        if (txtTodo.isEmpty()) {
            Snackbar
                .make(view, "오늘 할 일을 적어주세요!", Snackbar.LENGTH_SHORT)
                .show()
            return
        }

        var todoVO: TodoVO = TodoVO()
        if (btnAdd.text == "변경") {
            todoVO = btnAdd.tag as TodoVO
            todoVO.todo = txtTodo
            btnAdd.text = "추가"
        } else {
            val sd = SimpleDateFormat("yyyy-MM-dd")
            val st = SimpleDateFormat("HH:mm:ss")
            val date = System.currentTimeMillis()

            todoVO.todo = txtTodo
            todoVO.date = sd.format(date).toString()
            todoVO.time = st.format(date).toString()
        }
        todoViewModel.save(todoVO)
        txtTodoInput.text = null
    }

    fun todoUpdate(id: Long) {
        val todoVO: TodoVO = todoViewModel.findById(id)

        if (todoVO.isComplete) {
            Toast.makeText(this, "완료한 업무는 수정이 불가능합니다.", Toast.LENGTH_SHORT).show()
            return
        }
        txtTodoInput?.setText(todoVO.todo.toString())
        btnAdd.text = "변경"
        btnAdd.tag = todoVO
    }

    fun todoComplete(id: Long) {
        val todoVO: TodoVO = todoViewModel.findById(id)
        todoVO.isComplete = !todoVO.isComplete
        todoViewModel.save(todoVO)
    }
}
